text/0000-add-pack-interact.md

```diff
@@ -0,0 +1,83 @@
+# Meta
+[meta]: #meta
+- Name: Add Pack Interact
+- Start Date: 2020-04-05
+- Author(s): aemengo
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+[summary]: #summary
+
+Pack should provide an interactive mode for building images, updating dependencies, and other user workflows. Please see [mockup](https://storage.googleapis.com/artifacts.cf-garden-windows-dev.appspot.com/pack-interact-edit.mp4) for concept.
+
+# Definitions
+[definitions]: #definitions
+
+* [Pack](https://github.com/buildpacks/pack) - A tool within the buildpacks ecosystem for converting code into runnable images, in a local environment.
+* Build plan - A sequence of implementation buildpacks prescribed by the detect phase.
+
+# Motivation
+[motivation]: #motivation
+
+The **casual application developer** has trouble seeing the value proposition of buildpacks. They also encounter a high learning curve when using our tooling. We've seen evidence of this through user interviews and other channels. By providing a more interactive experience, the casual application developer will be more enticed to remain within the buildpacks ecosystem. Pack will act as the gateway tool to more advanced parts of the ecosystem.
+
+# What it is
+[what-it-is]: #what-it-is
+
+When a casual application developer runs `pack interact` on the command-line in the working directory of their application, they are provided with an interactive panel where they can perform activities related to converting code into runnable images. The following **2 user workflows** will be introduced in the interactive panel:
```

Comment:
One of the major things missing from this RFC is how this feature incorporate all the possible options allowed via `pack build`.

For example, how do I provide additional buildpacks, build-time env vars, etc. The full list of options below.

I wonder if an alternative would be to do something like `pack build -i` / `pack build --interact`. I understand that it might defeat some purpose in "interaction" but I think it would greatly simplify the maintenance of the feature.

```
Flags:
  -B, --builder string              Builder image (default "cnbs/sample-builder:bionic")
  -b, --buildpack strings           Buildpack to use. One of:
                                      a buildpack by id and version in the form of '<buildpack>@<version>',
                                      path to a buildpack directory (not supported on Windows),
                                      path/URL to a buildpack .tar or .tgz file, or
                                      a packaged buildpack image name in the form of '<hostname>/<repo>[:<tag>]'
                                    Repeat for each buildpack in order, or supply once by comma-separated list
  -r, --buildpack-registry string   Buildpack Registry by name
      --cache-image string          Cache build layers in remote registry. Requires --publish
      --clear-cache                 Clear image's associated cache before building
  -D, --default-process string      Set the default process type. (default "web")
  -d, --descriptor string           Path to the project descriptor file
      --docker-host string          Address to docker daemon that will be exposed to the build container.
                                    If not set (or set to empty string) the standard socket location will be used.
                                    Special value 'inherit' may be used in which case DOCKER_HOST environment variable will be used.
                                    This option may set DOCKER_HOST environment variable for the build container if needed.

  -e, --env stringArray             Build-time environment variable, in the form 'VAR=VALUE' or 'VAR'.
                                    When using latter value-less form, value will be taken from current
                                      environment at the time this command is executed.
                                    This flag may be specified multiple times and will override
                                      individual values defined by --env-file.
                                    Repeat for each env in order, or supply once by comma-separated list
                                    NOTE: These are NOT available at image runtime.
      --env-file stringArray        Build-time environment variables file
                                    One variable per line, of the form 'VAR=VALUE' or 'VAR'
                                    When using latter value-less form, value will be taken from current
                                      environment at the time this command is executed
                                    NOTE: These are NOT available at image runtime."
  -h, --help                        Help for 'build'
      --lifecycle-image string      Custom lifecycle image to use for analysis, restore, and export when builder is untrusted.
      --network string              Connect detect and build containers to network
  -p, --path string                 Path to app dir or zip-formatted file (defaults to current working directory)
      --publish                     Publish to registry
      --pull-policy string          Pull policy to use. Accepted values are always, never, and if-not-present. (default "always")
      --run-image string            Run image (defaults to default stack's run image)
  -t, --tag strings                 Additional tags to push the output image to.
                                    Repeat for each tag in order, or supply once by comma-separated list
      --trust-builder               Trust the provided builder
                                    All lifecycle phases will be run in a single container (if supported by the lifecycle).
      --volume stringArray          Mount host volume into the build container, in the form '<host path>:<target path>[:<options>]'.
                                    - 'host path': Name of the volume or absolute directory path to mount.
                                    - 'target path': The path where the file or directory is available in the container.
                                    - 'options' (default "ro"): An optional comma separated list of mount options.
                                        - "ro", volume contents are read-only.
                                        - "rw", volume contents are readable and writeable.
                                        - "volume-opt=<key>=<value>", can be specified more than once, takes a key-value pair consisting of the option name and its value.
                                    Repeat for each volume in order, or supply once by comma-separated list

Global Flags:
      --no-color     Disable color output
  -q, --quiet        Show less output
      --timestamps   Enable timestamps in output
  -v, --verbose      Show more output
```

Reply:
> how do I provide additional buildpacks, build-time env vars, etc

I would expect a user specify more details like that through the *(e) edit)* option. As of right now, I don't have a clear idea what it would look like - and I don't want to rush a solution for it. If we think it's out of scope for this RFC, then I'm fine with it.

Regardless, I don't want a user inundated with all the options at the "interactive panel".

> how about `pack build --interact`?

I don't have strong feelings here.

Reply:
@jromero do we need interact to support everything `build` vs `interact` at least initially?

I do wonder for env stuff if we push them to use `project.toml` instead?